internal/app/go-filecoin/paymentchannel/integration_test.go

```diff
@@ -54,25 +55,65 @@ func TestCreatePaymentChannel(t *testing.T) {
 
 	viewer := paychtest.NewFakeStateViewer(t)
 
-	pchMgr := pch.NewManager(context.Background(), ds, fms, fms, viewer)
+	pchMgr := pch.NewManager(context.Background(), ds, initActorUtil, initActorUtil, viewer)
 
 	viewer.GetFakeStateView().AddActorWithState(paych, client, miner, address.Undef)
 
 	rmc := retrievalmarketconnector.NewRetrievalMarketClientFakeAPI(t)
 
-	rcnc := retrievalmarketconnector.NewRetrievalClientConnector(bs, fakeProvider, rmc, pchMgr)
-	assert.NotNil(t, rcnc)
+	connector := retrievalmarketconnector.NewRetrievalClientConnector(bs, fakeProvider, rmc, pchMgr)
+	assert.NotNil(t, connector)
 	tok, err := encoding.Encode(genTs.Key())
 	require.NoError(t, err)
 
-	res, _, err := rcnc.GetOrCreatePaymentChannel(ctx, client, miner, channelAmt, tok)
+	addr, mcid, err := connector.GetOrCreatePaymentChannel(ctx, client, miner, initialChannelAmt, tok)
+	require.NoError(t, err)
+	assert.Equal(t, address.Undef, addr)
+
+	// let the goroutine finish creating channel info
+	time.Sleep(100 * time.Millisecond)
```

Comment:
The goroutine is called in GetOrCreatePaymentChannel; we want to return immediately the message CID to the caller but will still need to wait for the payment channel to be created, message to appear on chain and register the new payment channel info.

Reply:
Naively, it would seem like the `connector.WaitForPaymentChannelCreation` below should handle that, possibly combined with a message wait. If there's something else that also needs to be awaited, can we expose something  explicitly to await?

Reply:
I started to answer this and it got super long. I think it would be faster to resolve this in a Zoom chat.

Reply:
@anorth I have made some mutex adjustments that should be there anyway -- the payment Channel store should be locked completely until creation finishes or fails -- and LOTUS is setup this way. This removes the need for time.Sleep.

I agree that there are larger issues in the overall design of this module -- as documented in https://github.com/filecoin-project/go-filecoin/issues/4045 -- and I will work on an introductory technical problem statement. Even if we do make it a shared module, I don't think it should live in Markets, as my understanding is payment channels may eventually be used for other things besides retrieval?

Anyway, I'm wondering if we can move this forward for now, as separating the call from the waiting is important for the markets module, with the clear understanding that getting a better overall design for the payment channel manager and potentially making it a shared component is an important goal

Reply:
Yes, non-blocking. 👍 for landing it @shannonwells